web/wai-server/Main.hs

```diff
@@ -66,26 +67,35 @@ socketsApp ref = websocketsOr defaultConnectionOptions wsApp defaultApp where
     sendTextData conn (B.pack $ show i)
     putMVar convar (name, conn)
     forkPingThread conn 30
-    forever $ receiveMove i conn -- listen for moves, forever.
-
-  -- | receive a move (and game) from a player, and attempt to apply it
-  --   if it isn't the players turn, or the move results in an error,
-  --   then send the error message back on the connection
-  --   if it succeeds, send the new game state to both players
-  -- TODO: check if it is the players turn.
-  receiveMove :: Int -> Connection -> IO ()
-  receiveMove pid conn = go where
-    go = do
-      gameAndMove <- receiveData conn
-      case decodeGameAndMove gameAndMove of
-        Right g     -> updateBothClients g
+    forever $ receiveMessage i conn -- listen for moves, forever.
+
+-- | recieve a message from the client.
+--   decode the message and check if it's the player's turn
+--   propagating error message along the way
+--   then match on the MessageType and handle each message appropriately
+
+receiveMessage :: Int -> Connection -> IO ()
+receiveMessage pid conn = do
+    messageData <- receiveData conn
+    case decodeAndCheckTurn messageData of
+        Right (Message ActualMove g m) ->
+            case applyWordPut g m of --apply the move, update both clients
+                Right newGame -> updateBothClients newGame
+                Left errMsg -> sendTextData conn (B.pack errMsg)
```

Comment:
This is a nitpicky thing, but I like nicely aligned code. It makes you feel better when your code looks very nice. Can you do this:

     Right newGame -> updateBothClients newGame
     Left  errMsg  -> sendTextData conn (B.pack errMsg)